### DISCUSSION TITLE
How to get server address of TLS server

### DISCUSSION THREAD

**Discussion by JoelSatkas** [tokio-rs/axum#2329] [Q&A]
I followed the example of implementing a TLS server in axum: https://github.com/tokio-rs/axum/blob/main/examples/tls-graceful-shutdown/src/main.rs and adapted the code to my project but now i would like it that this https server asks the OS to just take the next available port number and then print to the console what this port number is so that its accessible.

This is possible to do for the http redirect server because it uses the hyper builder directly. This is how you can get axum to dynamically assign a port number adn then print to console:

```
// Create the socket address object from the given ip and port numbers.
let addr = SocketAddr::from_str(&(ip_address + ":" + ports.http.to_string().as_str())).unwrap();
// Create server and bind to created socket address, this will panic if unable to bind to ip address + port number.
// If given port number of 0, OS will assign an available port number.
let server_result =
    axum::Server::try_bind(&addr).expect(&format!("Failed to bind to address {}", addr));
// Creates server from builder, our main service is http redirection.
let server = server_result.serve(redirect.into_make_service());
// Only now can we access the OS assigned port number address.
tracing::info!("http redirect listening on {}", server.local_addr());

// Assign graceful shutdown signal and run the server consuming the thread until it shutdowns.
server
    .with_graceful_shutdown(shut_down_signal)
    .await
    .unwrap();

tracing::info!("http redirect shutdown");
```

But i also want to do this as well for the https server from which as i understand it, is not possible. And the given example would even be wrong if provided a port number of 0 because the OS would actually assign a different port number but 0 would be printed in this tracing log. 

```
// run https server
let addr = SocketAddr::from(([127, 0, 0, 1], ports.https));
tracing::debug!("listening on {addr}");
axum_server::bind_rustls(addr, config)
    .handle(handle)
    .serve(app.into_make_service())
    .await
    .unwrap();
```

Is it possible to do it at all?

**Comment by davidpdrsn** [comment]
`axum_server` is a third party crate so you need to ask there.

I have a look at their docs and indeed couldn't find a way to get the address of a [`Server`](https://docs.rs/axum-server/latest/axum_server/struct.Server.html).

**Reply by JoelSatkas** [reply]
A few minutes after posting the question i found the answer :D. I was looking through the axum_server code and saw a function `from_tcp_rustls` which is perfect because creating a tcp listener also binds to the address and the OS assigns it an available port number if it was given 0. So the solution is:

```
    let ip_and_port_for_https_server =
        axum_ip_address_to_use.clone() + ":" + ports.https.to_string().as_str();

    // Create listener for https server. If provided port number is 0 the OS will assign it an available port number.
    let listener = TcpListener::bind(&ip_and_port_for_https_server).expect(&format!(
        "Failed to bind https server to address {}",
        ip_and_port_for_https_server
    ));

    tracing::info!(
        "Axum https listening on {}",
        listener.local_addr().expect("Failed to get local address")
    );

    // Run the server. This will consume the thread until Axum is signaled to shutdown or it failed to launch.
    let server_launch_result: Result<(), std::io::Error> =
        axum_server::from_tcp_rustls(listener, rustls_config)
            .handle(handle)
            .serve(app.into_make_service())
            .await;
```

Seems to work for me. I should have spent some more time researching before asking but hopefully someone else might find this answer useful. Thanks for searching anyway.

**Reply by JoelSatkas** [reply]
Update: Turns out that asking the OS to assign you a port number (passing 0 as the port number) makes the code in the tls example behave differently. When making a request to the http server, instead of the code triggering and redirecting to the https server, there is now an error on the client side.

* In Chrome the error is: **This site can’t provide a secure connection [IP address] sent an invalid response.
ERR_SSL_PROTOCOL_ERROR.**
* In postman the error is: **Error: write EPROTO 77715720:error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER:../../../../src/third_party/boringssl/src/ssl/tls_record.cc:242**
* In Firefox the error is: **Secure Connection Failed. An error occurred during a connection to [IP address]. SSL received a record that exceeded the maximum permissible length. Error code: SSL_ERROR_RX_RECORD_TOO_LONG**

Each client seems to give a different response/reason for why it failed. Ive also noticed in Chrome that the server responds with a redirect 308 and gives the location of the same http url but just adds https as the protocol which is wrong. 

I dont know where this is coming from and its definitely not coming from the rust code as that doesnt even hit the break point. I dont know enough about networking to find a solution around this so i am forced to assign fixed port numbers because when doing so this problem does not occur.
